File: facebook/junitTests/src/test/java/com/facebook/login/LoginHandlerTestCase.java
/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * You are hereby granted a non-exclusive, worldwide, royalty-free license to use,
 * copy, modify, and distribute this software in source code or binary form for use
 * in connection with the web services and APIs provided by Facebook.
 *
 * As with any software that integrates with the Facebook platform, your use of
 * this software is subject to the Facebook Developer Principles and Policies
 * [http://developers.facebook.com/policy/]. This copyright notice shall be
 * included in all copies or substantial portions of the software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.facebook.login;

import android.app.Activity;
import android.support.v4.app.FragmentActivity;

import com.facebook.FacebookPowerMockTestCase;

import org.junit.Before;
import org.robolectric.Robolectric;

import java.util.Arrays;
import java.util.Date;
import java.util.HashSet;

import static org.junit.Assert.assertTrue;
import static org.powermock.api.mockito.PowerMockito.mock;
import static org.powermock.api.mockito.PowerMockito.when;

public abstract class LoginHandlerTestCase extends FacebookPowerMockTestCase {
    protected static final String ACCESS_TOKEN = "An access token";
    protected static final String USER_ID = "1000";
    protected static final long EXPIRES_IN_DELTA = 3600 * 24 * 60;
    protected static final HashSet<String> PERMISSIONS = new HashSet<String>(
            Arrays.asList("go outside", "come back in"));
    protected static final String ERROR_MESSAGE = "This is bad!";

    protected FragmentActivity activity;
    protected LoginClient mockLoginClient;

    @Before
    public void before() throws Exception {
        mockLoginClient = mock(LoginClient.class);
        activity = Robolectric.buildActivity(FragmentActivity.class).create().get();
        when(mockLoginClient.getActivity()).thenReturn(activity);
    }

    protected LoginClient.Request createRequest() {
        return createRequest(null);
    }

    protected LoginClient.Request createRequest(String previousAccessTokenString) {

        return new LoginClient.Request(
                LoginBehavior.SSO_WITH_FALLBACK,
                new HashSet<String>(PERMISSIONS),
                DefaultAudience.FRIENDS,
                "1234",
                "5678");
    }

    protected void assertDateDiffersWithinDelta(Date expected, Date actual, long expectedDifference,
                                                long deltaInMsec) {

        long delta = Math.abs(expected.getTime() - actual.getTime()) - expectedDifference;
        assertTrue(delta < deltaInMsec);
    }
}

File: facebook-common/src/test/kotlin/com/facebook/login/LoginHandlerTestCase.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.login

import androidx.fragment.app.FragmentActivity
import com.facebook.FacebookPowerMockTestCase
import com.facebook.login.AuthenticationTokenTestUtil.getEncodedAuthTokenStringForTest
import java.util.Date
import kotlin.math.abs
import org.assertj.core.api.Assertions.assertThat
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever
import org.robolectric.Robolectric

abstract class LoginHandlerTestCase : FacebookPowerMockTestCase() {
  protected lateinit var activity: FragmentActivity
  protected lateinit var mockLoginClient: LoginClient

  override fun setup() {
    super.setup()
    mockLoginClient = mock()
    activity = Robolectric.buildActivity(FragmentActivity::class.java).create().get()
    whenever(mockLoginClient.activity).thenReturn(activity)
  }

  @JvmOverloads
  protected open fun createRequest(previousAccessTokenString: String? = null): LoginClient.Request {
    return LoginClient.Request(
        LoginBehavior.NATIVE_WITH_FALLBACK,
        HashSet(PERMISSIONS),
        DefaultAudience.FRIENDS,
        "rerequest",
        "1234",
        "5678",
        LoginTargetApp.FACEBOOK,
        AuthenticationTokenTestUtil.NONCE,
        CODE_VERIFIER,
        CODE_CHALLENGE,
        CodeChallengeMethod.S256,
        "https://example.com/redirect/my_uri")
  }

  protected open fun createIGAppRequest(): LoginClient.Request {
    return LoginClient.Request(
        LoginBehavior.NATIVE_WITH_FALLBACK,
        HashSet(PERMISSIONS),
        DefaultAudience.FRIENDS,
        "rerequest",
        "1234",
        "5678",
        LoginTargetApp.INSTAGRAM)
  }

  protected open fun createIGWebRequest(): LoginClient.Request {
    return LoginClient.Request(
        LoginBehavior.WEB_ONLY,
        HashSet(PERMISSIONS),
        DefaultAudience.FRIENDS,
        "rerequest",
        "1234",
        "5678",
        LoginTargetApp.INSTAGRAM)
  }

  protected open fun createRequestWithNonce(): LoginClient.Request {
    val codeVerifier = PKCEUtil.generateCodeVerifier()
    return LoginClient.Request(
        LoginBehavior.NATIVE_WITH_FALLBACK,
        HashSet(PERMISSIONS),
        DefaultAudience.FRIENDS,
        "rerequest",
        "1234",
        "5678",
        null,
        AuthenticationTokenTestUtil.NONCE,
        codeVerifier,
        PKCEUtil.generateCodeChallenge(codeVerifier, CodeChallengeMethod.S256),
        CodeChallengeMethod.S256)
  }

  protected fun assertDateDiffersWithinDelta(
      expected: Date,
      actual: Date,
      expectedDifference: Long,
      deltaInMsec: Long
  ) {
    val delta = abs(expected.time - actual.time) - expectedDifference
    assertThat(delta).isLessThan(deltaInMsec)
  }

  protected val encodedAuthTokenString: String
    get() = getEncodedAuthTokenStringForTest()

  companion object {
    const val ACCESS_TOKEN = "An access token"
    const val USER_ID = "123"
    const val EXPIRES_IN_DELTA = (3600 * 24 * 60).toLong()
    @JvmField val PERMISSIONS = setOf("go outside", "come back in")
    const val ERROR_MESSAGE = "This is bad!"
    const val CODE_VERIFIER = "codeVerifier"
    const val CODE_CHALLENGE = "codeChallenge"
  }
}
